activemodel/lib/active_model/normalization.rb

```diff
@@ -0,0 +1,156 @@
+# frozen_string_literal: true
+
+module ActiveModel # :nodoc:
+  module Normalization
+    extend ActiveSupport::Concern
+
+    included do
+      include ActiveModel::Dirty
+      include ActiveModel::Validations::Callbacks
+
+      class_attribute :normalized_attributes, default: Set.new
+
+      before_validation :normalize_changed_in_place_attributes
+    end
+
+    # Normalizes a specified attribute using its declared normalizations.
+    #
+    # ==== Examples
+    #
+    #   class User
+    #     include ActiveModel::Model
+    #     include ActiveModel::Attributes
+    #     include ActiveModel::Normalization
+    #
+    #     attribute :email, :string
+    #
+    #     normalizes :email, with: -> email { email.strip.downcase }
+    #   end
+    #
+    #   legacy_user = User.find(1)
+    #   legacy_user.email # => " CRUISE-CONTROL@EXAMPLE.COM\n"
+    #   legacy_user.normalize_attribute(:email)
+    #   legacy_user.email # => "cruise-control@example.com"
+    def normalize_attribute(name)
+      # Treat the value as a new, unnormalized value.
+      send(:"#{name}=", send(name))
+    end
+
+    module ClassMethods
+      # Declares a normalization for one or more attributes. The normalization
+      # is applied when the attribute is assigned or validated.
+      #
+      # Because the normalization may be applied multiple times, it should be
+      # _idempotent_. In other words, applying the normalization more than once
+      # should have the same result as applying it only once.
+      #
+      # By default, the normalization will not be applied to +nil+ values. This
+      # behavior can be changed with the +:apply_to_nil+ option.
+      #
+      # ==== Options
+      #
+      # * +:with+ - Any callable object that accepts the attribute's value as
+      #   its sole argument, and returns it normalized.
+      # * +:apply_to_nil+ - Whether to apply the normalization to +nil+ values.
+      #   Defaults to +false+.
+      #
+      # ==== Examples
+      #
+      #   class User
+      #     include ActiveModel::Model
+      #     include ActiveModel::Attributes
+      #     include ActiveModel::Normalization
+      #
+      #     attribute :email, :string
+      #     attribute :phone, :string
+      #
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #     normalizes :phone, with: -> phone { phone.delete("^0-9").delete_prefix("1") }
+      #   end
+      #
+      #   user = User.new(email: " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   user.email                  # => "cruise-control@example.com"
```

Comment:
The whitespace preceding `# =>` was to match the indentation of [subsequent lines](https://github.com/rails/rails/blob/f8eba0c047fa3976d7af24e194c99125505a3b79/activerecord/lib/active_record/normalization.rb#L74-L79).  No need for it here.